pkg/controllers/mcs/multiclusterservice_controller.go

```diff
@@ -0,0 +1,620 @@
+		err := c.Client.Get(context.TODO(), types.NamespacedName{Namespace: fedKey.Namespace, Name: fedKey.Name}, svcExport)
+		if err != nil {
+			return nil, err
+		}
+		return helper.ToUnstructured(svcExport)
+	}
+	return nil, fmt.Errorf("unsupported kind(%s)", fedKey.Kind)
+}
+
+func (c *MultiClusterServiceController) newMultiClusterServiceFunc() handler.MapFunc {
+	return func(object client.Object) (requests []reconcile.Request) {
+		mcs := &networkingv1alpha1.MultiClusterService{}
+		switch t := object.(type) {
+		case *workv1alpha2.ResourceBinding:
+			err := c.Client.Get(context.TODO(), types.NamespacedName{Namespace: t.Spec.Resource.Namespace, Name: t.Spec.Resource.Name}, mcs)
+			if err != nil {
+				return nil
+			}
+			requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+				Namespace: mcs.Namespace,
+				Name:      mcs.Name,
+			}})
+		case *clusterv1alpha1.Cluster:
+			mcsLists := &networkingv1alpha1.MultiClusterServiceList{}
+			err := c.Client.List(context.TODO(), mcsLists)
+			if err != nil {
+				return nil
+			}
+			for _, item := range mcsLists.Items {
+				if len(item.Spec.Range.ClusterNames) == 0 {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+						Namespace: item.Namespace,
+						Name:      item.Name,
+					}})
+					continue
+				}
+				for _, clusterName := range item.Spec.Range.ClusterNames {
+					if clusterName != t.Name {
+						continue
+					}
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+						Namespace: item.Namespace,
+						Name:      item.Name,
+					}})
+				}
+			}
+		default:
+			err := c.Client.Get(context.TODO(), types.NamespacedName{Namespace: t.GetNamespace(), Name: t.GetName()}, mcs)
+			if err != nil {
+				return nil
+			}
+			requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+				Namespace: mcs.Namespace,
+				Name:      mcs.Name,
+			}})
+		}
+		return
+	}
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *MultiClusterServiceController) SetupWithManager(mgr controllerruntime.Manager) error {
+	return utilerrors.NewAggregate([]error{
+		mgr.Add(c),
+		controllerruntime.NewControllerManagedBy(mgr).
+			For(&networkingv1alpha1.MultiClusterService{}).
+			Watches(&source.Kind{Type: &corev1.Service{}}, handler.EnqueueRequestsFromMapFunc(c.newMultiClusterServiceFunc())).
+			Watches(&source.Kind{Type: &workv1alpha2.ResourceBinding{}}, handler.EnqueueRequestsFromMapFunc(c.newMultiClusterServiceFunc()),
+				// Only observe the rb of the svc.
+				builder.WithPredicates(predicate.Funcs{
+					CreateFunc: func(createEvent event.CreateEvent) bool {
+						newRB := createEvent.Object.(*workv1alpha2.ResourceBinding)
+						if newRB.Spec.Resource.APIVersion != corev1.SchemeGroupVersion.String() || newRB.Spec.Resource.Kind != util.ServiceKind {
+							return false
+						}
+						return true
```

Comment:
Can be replaced with:
```go
return newRB.Spec.Resource.Kind == util.ServiceKind && newRB.Spec.Resource.APIVersion == corev1.SchemeGroupVersion.String()
```

Besides, I'd like to discuss whether we can ignore judgment APIVersion. Since Kubernetes v1.0, the Service version is v1.